### PR TITLE
[Merged by Bors] - feat(RingTheory/AdicCompletion): functoriality

### PR DESCRIPTION
**Pull request by chrisflav**
Adds functoriality of adic completions.

---

- [x] depends on: #12516
- [x] depends on: #12553 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (erdOne) [inline Mathlib/RingTheory/AdicCompletion/Functoriality.lean:55]: Why `R ⧸ (I • ⊤ : Ideal R)` instead of `R ⧸ I`?

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (erdOne) [inline Mathlib/RingTheory/AdicCompletion/Functoriality.lean:76]: I think this should be called `AdicCauchySequence.map`?

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (erdOne) [inline Mathlib/RingTheory/AdicCompletion/Functoriality.lean:118]: Similar for this.

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (erdOne) [inline Mathlib/RingTheory/AdicCompletion/Functoriality.lean:133]: Is this intended to be called `LinearMap.AdicCompletion.ext`?

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (erdOne) [inline Mathlib/RingTheory/AdicCompletion/Functoriality.lean:145]: Is `AdicCompletion.mk` a linearMap? If so, maybe `(h : f.comp (AdicCompletion.mk I M a) = _)` instead.  See note [partially-applied ext lemmas]

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (chrisflav) [reply Mathlib/RingTheory/AdicCompletion/Functoriality.lean]: I changed it to `AdicCompletion.map_ext`

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (chrisflav) [reply Mathlib/RingTheory/AdicCompletion/Functoriality.lean]: Yes it is, I changed it. For the `map_ext''' version, I left it as is for now, because I did not see a way without asking more conditions on `T`. If this is desirable, let me know.

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (chrisflav) [reply Mathlib/RingTheory/AdicCompletion/Functoriality.lean]: Because I thought it is more convenient to work with, because of the definition of `AdicCompletion I R`. Turns out, it's also doable with the canonical spelling, so I changed it.

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (erdOne) [inline Mathlib/RingTheory/AdicCompletion/Functoriality.lean:267]:
```suggestion
  letI f := map I (linearEquivFunOnFintype R ι M)
  letI g := linearEquivFunOnFintype (AdicCompletion I R) ι (fun j ↦ AdicCompletion I (M j))
  g.symm.toLinearMap ∘ₗ pi I M ∘ₗ f
```
This makes `let` disappear in the constructed term.

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (erdOne) [inline Mathlib/RingTheory/AdicCompletion/Functoriality.lean:44]:
```suggestion
@[local simp]
```

Comment (erdOne) [reply Mathlib/RingTheory/AdicCompletion/Functoriality.lean:44]: Did you really use this somewhere? I doubt so.

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (erdOne) [inline Mathlib/RingTheory/AdicCompletion/Functoriality.lean:116]:
```suggestion
@[local simp]
```
I don't think you used this anywhere as well?

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (erdOne) [inline Mathlib/RingTheory/AdicCompletion/Functoriality.lean:330]:
```suggestion
  letI f := (congr I (linearEquivFunOnFintype R ι M)).symm
  letI g := (linearEquivFunOnFintype (AdicCompletion I R) ι (fun j ↦ AdicCompletion I (M j)))
```

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (chrisflav) [reply Mathlib/RingTheory/AdicCompletion/Functoriality.lean]: No, they only appear in `simp only` calls. I changed it to `local simp` as suggested, should I remove it completely instead?

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (erdOne) [inline Mathlib/RingTheory/AdicCompletion/Functoriality.lean:146]: This should be the `@[ext]` lemma instead. See [partially-applied ext lemmas].

**Review by erdOne — APPROVED** [review]
Sorry forgot about this.
LGTM modulo the comment

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (chrisflav) [reply Mathlib/RingTheory/AdicCompletion/Functoriality.lean:146]: Instead of `map_ext`?

Comment (erdOne) [reply Mathlib/RingTheory/AdicCompletion/Functoriality.lean:146]: Yes.

Comment (erdOne) [reply Mathlib/RingTheory/AdicCompletion/Functoriality.lean:146]: Okay maybe it doesn't matter.

**Comment by erdOne** [comment]
Thanks!
maintainer merge

**Comment by chrisflav** [comment]
Thanks for the review @erdOne!

File: Mathlib/Algebra/DirectSum/Basic.lean
Comment (riccardobrasca) [inline Mathlib/Algebra/DirectSum/Basic.lean:155]: Do we really need this `change`? If so we are maybe missing a lemma.

File: Mathlib/RingTheory/AdicCompletion/Algebra.lean
Comment (riccardobrasca) [inline Mathlib/RingTheory/AdicCompletion/Algebra.lean:161]: Nonterminal `simp`.

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (riccardobrasca) [inline Mathlib/RingTheory/AdicCompletion/Functoriality.lean:18]:
```suggestion
- `LinearMap.adicCauchy I f`: the linear map on `I`-adic cauchy sequences induced by `f`
- `LinearMap.adicCompletion I f`: the by `f` induced linear map on `I`-adic completions
```
Is this what you mean?

File: Mathlib/Algebra/DirectSum/Basic.lean
Comment (chrisflav) [reply Mathlib/Algebra/DirectSum/Basic.lean:155]: I tried to add various rewrite lemmas, even `lemma foo [Fintype ι] (x : ⨁ i, β i) : /∑ i ∈ Finset.univ, of β i (x i)) = ∑ i ∈ Finset.univ, DFinsupp.single i (x i)`, but it I can't get it to work without the `change`.

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (chrisflav) [reply Mathlib/RingTheory/AdicCompletion/Functoriality.lean]: Yes

File: Mathlib/RingTheory/AdicCompletion/Functoriality.lean
Comment (chrisflav) [reply Mathlib/RingTheory/AdicCompletion/Functoriality.lean]: I swapped the `@[ext]`.

File: Mathlib/RingTheory/AdicCompletion/Algebra.lean
Comment (riccardobrasca) [inline Mathlib/RingTheory/AdicCompletion/Algebra.lean:161]: Can you use `dsimp only`? Thanks!

**Review by riccardobrasca — COMMENTED** [review]
Thanks!

bors d+

**Comment by chrisflav** [comment]
Thanks for the reviews!

bors r+

**Pull request closed without merging**
